dap_aria_mapping/pipeline/make_topic_name_assignments.py

```diff
@@ -0,0 +1,432 @@
+
+                            # Attempt to convert to list, if exception, try making it explicit
+                            try:
+                                response = ast.literal_eval(response)
+                            except Exception as e:
+                                logger.info(
+                                    f"FAILURE - ChatGPT response is not a list: {response}."
+                                )
+                                sleep_time = np.random.randint(4, 6)
+                                logger.info(
+                                    f"Routine idling - Sleeping for {sleep_time} seconds"
+                                )
+                                time.sleep(sleep_time)
+                                for data in chatbot.ask(query):
+                                    response = data["message"]
+
+                            # In case it failed to output a list the first time but not the second
+                            try:
+                                if not isinstance(response, list):
+                                    response = ast.literal_eval(response)
+                            except:
+                                raise Exception("ChatGPT response is not a list.")
+
+                            logger.info(f"SUCCESS - ChatGPT response: {response}")
+                            for quadtuple in response:
+                                list_id = quadtuple[0].split(" ")[-1]
+                                topic = quadtuple[1]
+                                confidence = quadtuple[2]
+                                discard = quadtuple[3]
+                                chatgpt_names[list_id] = {
+                                    "name": topic,
+                                    "confidence": confidence,
+                                    "discard": discard,
+                                }
+
+                            # Accept response, refresh session, set to skip problem description
+                            response_ok = True
+                            if first[f"chatbot{chatbot_num}"]:
+                                first[f"chatbot{chatbot_num}"] = False
+
+                            # Update number of topics with names & pending groups
+                            num_topics_file = len(chatgpt_names.keys())
+                            entity_names = {
+                                k: v
+                                for k, v in entity_names.items()
+                                if k not in chatgpt_names.keys()
+                            }
+
+                            sleep_time = np.random.randint(9, 12)
+                            logger.info(
+                                f"Routine idling - Sleeping for {sleep_time} seconds"
+                            )
+                            time.sleep(sleep_time)
+
+                            # load back (in case other streams have updated the dictionary)
```

Comment:
Could you use pooling from the multiprocessing library to send different chunks of the entities to different workers, and then combine the output at the end, rather than worrying about having different workers updating the same dictionary?

Reply:
In the case of the reverse bot, I wrote it in a way that guarantees minimum work loss in case anything goes awry. Internet disconnections, simultaneous writting attempts, etc. I could probably have written it so that on error it dumps results recorded so far (so that you can parallelize & not lose anything).

This part of the code corresponds to the other bot: the problem here is that you need to manually log in via the simulated browser to get through chat gpt's api. These sessions are kept alive by an xserver, and if I had different threads working with different browsers, it would crash. But yeah, if possible your approach would be best.